### PR TITLE
Major optimizations to recipe tree traversal

### PR DESCRIPTION
**Pull request by planetguy32**
New code takes around .336 sec to show craftable items whereas old code took 12.830 sec for the same player inventory. ~30 mods installed, 1697 recipes available.

### CODE REVIEW

**Comment by planetguy32** [comment]
I'm probably going to be able to get rid of another foreach over all recipes in getting valid recipes, but I haven't right this moment.

**Comment by planetguy32** [comment]
Moving from my development workspace to a big modpack, it seems to hand in an infinite loop. I think I might be mis-handling the plugins, but I'd recommend not pulling my patches just yet.

**Comment by makotech222** [comment]
Okay thanks. Haven't had time to look at it yet, but i could see that the recipe searching could've done with a few optimizations :). Good job on it. Will pull when you give the go ahead.

**Comment by planetguy32** [comment]
The loop doesn't happen with the last build from Lepko, but when I checked out the code from before my patches it happened.

I can reproduce it using only Tinker's Construct and Easy Crafting with your patches.

**Comment by makotech222** [comment]
Interesting. I wasn't aware any of my changes crashed the game. Care to take a look at it for me? I don't have my java dev setup at the moment, and I don't have the time to look at it until at least the weekend.

**Comment by planetguy32** [comment]
It doesn't crash the game, it just hangs forever trying to list craftable items.

I'll have a look, but I'm not sure how much time I can put into it either.

**Comment by planetguy32** [comment]
I think TCon's wool recolouring recipes cause trouble, since stopping it for debug inspection shows that many of the recipes are crafting wool with dyes. If recursion is 10 and each recipe can be made with the results of each other recipe, that's 16^10 recipes to search through - ouch!

**Comment by planetguy32** [comment]
My thoughts on how to resolve this would be to mark recipes that use the same item as their outputs as one of their inputs as potentially pathological cases, and only follow one level deep of recursion.

**Comment by makotech222** [comment]
That would be a good idea to do. I would also like to add in a Map to store recipes that have already been confirmed craftable/uncraftable with the current inventory. Should speed up a lot as well. I'm gonna plan on trying these additions this weekend. Feel free to get there before me :)

Is this ready to pull?

**Comment by planetguy32** [comment]
Yes.

File: src/main/java/net/lepko/easycrafting/core/recipe/RecipeManager.java [hunk 1]
```diff
@@ -76,8 +83,13 @@ public static void scanRecipes() {
 		HashMap<String,String> recipeExists = new HashMap<String,String>();
 		for (IRecipe r : recipes) {
 			WrappedRecipe wr = WrappedRecipe.of(r);
-			if (wr != null && !CheckIfRecipeAlreadyExists(allRecipes, wr,recipeExists))
-			{
+			if (wr != null && !CheckIfRecipeAlreadyExists(allRecipes, wr,recipeExists)){
```
Comment (makotech222) [inline src/main/java/net/lepko/easycrafting/core/recipe/RecipeManager.java:86]: If i comment these lines out, my search finishes fine. Not sure what the purpose of it is, haven't looked too deep into it.
lines 87-92

**Pull request merged**